File: fifteen-exercises/add-two-numbers.rs
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

impl Solution {
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut l1 = l1.as_ref();
        let mut l2 = l2.as_ref();
        let mut head = Box::new(ListNode::new(0));
        let mut cur = &mut head;
        let mut carry = 0i32;
        while l1.is_some() || l2.is_some() || carry != 0 {
            // 对Option使用map直接处理Some可真是方便
            let sum = l1.map_or(0, |l| l.val) + l2.map_or(0, |l| l.val) + carry;
            carry = sum / 10;
            cur.next = Some(Box::new(ListNode::new(sum % 10)));
            cur = cur.next.as_mut().unwrap();
            l1 = l1.map(|l| l.next.as_ref()).unwrap_or(None);
            l2 = l2.map(|l| l.next.as_ref()).unwrap_or(None);
        }
        head.next
    }
}

struct Solution;

fn main() {
    // 比较麻烦，直接用OJ测试了
    println!("ok")
}

File: fifteen-exercises/add-two-numbers.cpp
#include <iostream>

struct ListNode {
    int val;
    ListNode *next;

    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        int carry = 0;
        auto *head = new ListNode(0);
        ListNode *cur = head;
        while (l1 != nullptr or l2 != nullptr or carry) {
            int sum = (l1 != nullptr ? l1->val : 0) + (l2 != nullptr ? l2->val : 0) + carry;
            carry = sum / 10;
            cur->next = new ListNode(sum % 10);
            cur = cur->next;
            l1 = l1 != nullptr ? l1->next : nullptr;
            l2 = l2 != nullptr ? l2->next : nullptr;
        }
        return head->next;
    }
};

int main() {
    auto *l1 = new ListNode(2);
    l1->next = new ListNode(4);
    l1->next->next = new ListNode(3);
    auto *l2 = new ListNode(5);
    l2->next = new ListNode(6);
    l2->next->next = new ListNode(4);

    Solution sln = Solution();
    ListNode *res = sln.addTwoNumbers(l1, l2);
    std::cout << res->val << std::endl;
    std::cout << res->next->val << std::endl;
    std::cout << res->next->next->val << std::endl;
    return 0;
}